### ISSUE TITLE
FPGA Compile in Vivado 2018.3

### ISSUE THREAD

**Issue by xjiang3** [pulp-platform/pulpino#270]
Hi,
I encountered a problem when I used   **make all**   in folder /pulpino/fpga to synthesize in Vivado 2018.3. Does anyone know how to fix this problem? 
Thanks,

```
# update_compile_order -fileset sources_1
# update_compile_order -fileset sim_1
# get_property source_mgmt_mode [current_project]
# set_property source_mgmt_mode DisplayOnly [current_project]
# get_property source_mgmt_mode [current_project]
# set_property strategy Flow_AreaOptimized_High [get_runs synth_1]
# synth_design -rtl -name rtl_1
Command: synth_design -rtl -name rtl_1
Starting synth_design
Using part: xc7z020clg484-1
Top: pulpemu_top
---------------------------------------------------------------------------------
Starting RTL Elaboration : Time (s): cpu = 00:00:01 ; elapsed = 00:00:01 . Memory (MB): peak = 1810.840 ; gain = 0.000 ; free physical = 11616 ; free virtual = 45132
---------------------------------------------------------------------------------
INFO: [Synth 8-6157] synthesizing module 'pulpemu_top' [/home/jiangxun/github/pulpino/fpga/rtl/pulpemu_top.v:11]
INFO: [Synth 8-6157] synthesizing module 'ps7_wrapper' [/home/jiangxun/github/pulpino/fpga/pulpemu/pulpemu.srcs/sources_1/bd/ps7/hdl/ps7_wrapper.v:12]
INFO: [Synth 8-6157] synthesizing module 'ps7' [/home/jiangxun/github/pulpino/fpga/pulpemu/pulpemu.srcs/sources_1/bd/ps7/synth/ps7.v:13]
ERROR: [Synth 8-439] module 'ps7_axi_crossbar_0_0' not found [/home/jiangxun/github/pulpino/fpga/pulpemu/pulpemu.srcs/sources_1/bd/ps7/synth/ps7.v:486]
ERROR: [Synth 8-6156] failed synthesizing module 'ps7' [/home/jiangxun/github/pulpino/fpga/pulpemu/pulpemu.srcs/sources_1/bd/ps7/synth/ps7.v:13]
ERROR: [Synth 8-6156] failed synthesizing module 'ps7_wrapper' [/home/jiangxun/github/pulpino/fpga/pulpemu/pulpemu.srcs/sources_1/bd/ps7/hdl/ps7_wrapper.v:12]
ERROR: [Synth 8-6156] failed synthesizing module 'pulpemu_top' [/home/jiangxun/github/pulpino/fpga/rtl/pulpemu_top.v:11]
---------------------------------------------------------------------------------
Finished RTL Elaboration : Time (s): cpu = 00:00:02 ; elapsed = 00:00:02 . Memory (MB): peak = 1810.840 ; gain = 0.000 ; free physical = 11625 ; free virtual = 45141
---------------------------------------------------------------------------------
RTL Elaboration failed
3 Infos, 0 Warnings, 0 Critical Warnings and 5 Errors encountered.
synth_design failed
ERROR: [Vivado_Tcl 4-5] Elaboration failed - please see the console for details
INFO: [Common 17-206] Exiting Vivado at Sat Jan  5 16:18:47 2019...
Makefile:6: recipe for target 'synth' failed
make[1]: *** [synth] Error 1
make[1]: Leaving directory '/home/jiangxun/github/pulpino/fpga/pulpemu'
Makefile:46: recipe for target 'pulpemu/pulpemu_top.bit' failed
make: *** [pulpemu/pulpemu_top.bit] Error 2
```

**Comment by stefanct** [comment]
I am also trying to get synthesis to work in 2018.3 and end up with the same error. One way to get a bit further is to simply remove/comment out the `synth_design -rtl -name rtl_1` call in `fpga/pulpemu/tcl/run.tcl`. I am not exactly sure about the consequences but synthesis completes then with no obvious problems. However, implementation later fails with some errors that some reported in the [xilinx forums](https://forums.xilinx.com/t5/Implementation/Place-design-issue-ERROR-DRC-REQP-127/td-p/809768) (without too much knowledge apparently).
The diagnosis of the Xilinx guy is that there is a bug in the design. For my naive newbie eyes it certainly looks like something is badly broken somewhere but I don't understand why it works fine with Vivado 2015.1 then. The resulting errors are:

    ERROR: [DRC REQP-127] obuf_loaded: OBUF pulpino_wrap_i/gpio_out_OBUF[16]_inst pin O drives one or more invalid loads. The loads are: ps7_wrapper_i/ps7_i/axi_gpio_emu/U0/gpio_core_1/Not_Dual.INPUT_DOUBLE_REGS3/GENERATE_LEVEL_P_S_CDC.MULTI_BIT.FOR_IN_cdc_to[16].CROSS2_PLEVEL_IN2SCNDRY_IN_cdc_to 
    ERROR: [DRC REQP-127] obuf_loaded: OBUF pulpino_wrap_i/gpio_out_OBUF[17]_inst pin O drives one or more invalid loads. The loads are: ps7_wrapper_i/ps7_i/axi_gpio_emu/U0/gpio_core_1/Not_Dual.INPUT_DOUBLE_REGS3/GENERATE_LEVEL_P_S_CDC.MULTI_BIT.FOR_IN_cdc_to[17].CROSS2_PLEVEL_IN2SCNDRY_IN_cdc_to 
    ERROR: [DRC REQP-127] obuf_loaded: OBUF pulpino_wrap_i/tdo_o_OBUF_inst pin O drives one or more invalid loads. The loads are: ps7_wrapper_i/ps7_i/axi_jtag_emu/U0/gpio_core_1/Not_Dual.INPUT_DOUBLE_REGS3/GENERATE_LEVEL_P_S_CDC.MULTI_BIT.FOR_IN_cdc_to[4].CROSS2_PLEVEL_IN2SCNDRY_IN_cdc_to

**Comment by FrancescoConti** [comment]
@stefanct the problem you mentioned seems related to the fact that the synthesis is split in two blocks (`pulpino` and `pulpemu`, which wraps it), without removing input and output buffers (`IBUF`/`OBUF`). I am quite sure that Vivado 2015.1 and older removed these buffers automatically when importing the `pulpino` netlist. This seems not the case any more in 2018.3.

Regarding the `synth_design -rtl -name rtl_1` command, this is (was?) used to perform RTL elaboration. The reason it fails seems to be related to something wrong with the block design instantiating the PS and a few Xilinx PL IPs. This may very well be non-working when you change the Vivado version; in fact I'd be more surprised if it worked out of the box. Why it seems to work without this operation escapes me, but perhaps the tools have changed a bit in the past few years.

**Comment by stefanct** [comment]
Yes, I was also thinking that they might have been optimized away with earlier versions and I think the majority still are but some remain for unknown reasons. Any ideas on approaches to get around that are very welcome ;)

Regarding the BD I have no theory at all... as suggested by Xilinx I have run the creation tcl with the original version (2015.1), opened up the resulting project with 2018.3 and let it write a new version with `write_bd_tcl`. Comparing the two scripts did not show any significant changes but I have nevertheless applied them in my update branch (after manually incorporating the switch for the Zybo board). However, that does not change anything regarding the `ps7_axi_crossbar_0_0` problem. Also, it looks quite fine to me in the GUI. ¯\_(ツ)_/¯

**Comment by matassos** [comment]
Hello i am meeting the same error you described above https://github.com/pulp-platform/pulpino/issues/270#issuecomment-463346089 did you find a solution?

**Comment by saahm** [comment]
@stefanct I tried using your fork at https://github.com/stefanct/pulpino/tree/xilinx_2018.3_tcl but it doesnt let me update the ips because the update-ips.py script references relatively from the repo it was called from basically (thorws me `fatal: repository 'https://github.com/stefanct/IPApproX.git/' not found`). And since you did not fork the IPApproX repo it cannot update the ips.

**Comment by matassos** [comment]
By the way, I found a solution on the above issue. 
As @FrancescoConti mentioned Vivado keeps the input and output buffers of the pulpino project.
Open the pulpino project on the new Vivado, I assume you have passed the steps of updating the ips of the project.
Update synthesis settings to "2018" and write "-mode out_of_context" on the more options entry.
After running the synthesis you can see on the schematic tab if there are buffers or not.

**Comment by stefanct** [comment]
On Mon, 06 May 2019 07:34:51 +0000 (UTC)
saahm <notifications@github.com> wrote:

> @stefanct I tried using your fork at https://github.com/stefanct/pulpino/tree/xilinx_2018.3_tcl but it doesnt let me update the ips because the update-ips.py script references relatively from the repo it was called from basically (thorws me `fatal: repository 'https://github.com/stefanct/IPApproX.git/' not found`). And since you did not fork the IPApproX repo it cannot update the ips. 

@saahm Did you make a completely new clone of my repo or did you add it
as a new remote to your existing pulpino repo that has upstream as
"origin" remote and simply fetched and "checkouted" that branch? 

I think the latter should work fine but you are right and doing a
complete clone does not. You could hack around that in the python
scripts but I did not. This is one of the problems coming from the
non-standard way of Pulpino to handle external repositories.

-- 
Kind regards/Mit freundlichen Grüßen, Stefan Tauner

**Comment by saahm** [comment]
Oh I see, yes I did only clone your fork and did not check it out as an upstream. I'm fairly new to that features of git. You mind explaining how to properly check out your branch or pointing me towards necessary resources?

**Comment by stefanct** [comment]
@saahm, basically what I wrote in prose above. From the top of my head it should be approximately this:

    cd pulpino
    git remote add stefanct https://github.com/stefanct/pulpino
    git fetch stefanct
    git checkout -b xilinx_2018.3_tcl stefanct/xilinx_2018.3_tcl

**Comment by saahm** [comment]
Thanks. This seemed to check out your branch into the repository on my machine. 
But sadly it wont fully compile here too. There seems to be no rule to make `„ips/xilinx_mem_32768x32_dp/ip/xilinx_mem_32768x32_dp.dcp“`for `„pulpino/pulpino.edn“`. 
Interestingly on the checked out repo I worked on this memory ip was available from what I see in the directories.

**Comment by stefanct** [comment]
> Thanks. This seemed to check out your branch into the repository on my machine.
> But sadly it wont fully compile here too. There seems to be no rule to make `„ips/xilinx_mem_32768x32_dp/ip/xilinx_mem_32768x32_dp.dcp“`for `„pulpino/pulpino.edn“`.
> Interestingly on the checked out repo I worked on this memory ip was available from what I see in the directories.

Of course not: I removed them since they are unused/useless in d5d383b. Did you run the
`update-ips.py` script after checking my branch out as discussed
earlier?

**Comment by saahm** [comment]
Yes, I did your steps and then in the fpga directory
```
    make clean
    cd ..
    ./update-ips.py
    cd fpga
    make all
```
to make sure everything is properly setup before make. From the other issue I only saw that you removed the 8k ips.

**Comment by saahm** [comment]
@stefanct I have also tried cloning a clean repository from the pulpino repo and repeat your steps to make sure there was nothing leftover of edits by me. 
So on the first one where I tried following the steps i built upon the repo i made slight edits too, tho I used git stash before fetching your repository into it. In there it causes the error with the 32k memory ip. 
On the cleanly cloned repository I did no edits, ran your fetch commands you gave me, and called the update-ips script aswell. There I get the error for the axi-crossbar I mentioned already the other issue. This happens when building pulpemu/pulpemu_top. Yet I dont get why it says it cannot find the ps7_axi_crossbar_0_0 because it is in `pulpemu/pulpemu.srcs/sources_1/bd/ps7/ip/` -- which is generated by the scripts and makefiles. 

Any idea what I could try to get this building fully? Do I simply dont see what I do wrong maybe?

**Comment by AlvinRolling** [comment]
Hi, @xjiang3 @stefanct , I got some answers for the above two errors. 

1. I address the first error by adding the ps7.bd file into the projects.
`add_files ./pulpemu.srcs/sources_1/bd/ps7/ps7.bd`

2. The second error seems quite weird. Notice that the OBUF of LD_o is automatically removed, since there is an extra register between gpio_out [15:8] and the output LD_o. oled_dc_o and oled_res_o, however, directly connects to the gpio_out[16] and [17], and thus caused this error. I did some bottom-up design method tests using simple RTL codes, and Vivado did juse removed the OBUF automatilly no matter whether or not there is a register between the output and the submodule. This is weird, and I still could figure out why. But by just adding a register between the gpio_out[16] & [17] and the outputs, this error could be addressed. 

> assign oled_dc_o = oled_dc_o_q;
always @(posedge s_clk_pulpino or negedge s_rstn_pulpino)
begin
   if (~s_rstn_pulpino)
      oled_dc_o_q <= 1'b0;
    else
      oled_dc_o_q <= gpio_out[16];
  end

**Comment by stefanct** [comment]
@AlvinRolling where did you add the `add_files` command exactly? I get a warning and it doesn't help when I add it after `add_files […] ps7_wrapper.v` in `run.tcl`.
The warning is `CRITICAL WARNING: [Vivado 12-1462] The source file is already part of the fileset 'sources_1'. Requested source '[…]/fpga/pulpemu/pulpemu.srcs/sources_1/bd/ps7/ps7.bd' will not be added.`

**Comment by stefanct** [comment]
Implementing the second suggestion of @AlvinRolling (and skipping the elaboration as explained earlier) another problem pops up: `ERROR: [DRC REQP-127] obuf_loaded: OBUF pulpino_wrap_i/tdo_o_OBUF_inst pin O drives one or more invalid loads. The loads are: ps7_wrapper_i/ps7_i/axi_jtag_emu/U0/gpio_core_1/Not_Dual.INPUT_DOUBLE_REGS3/GENERATE_LEVEL_P_S_CDC.MULTI_BIT.FOR_IN_cdc_to[4].CROSS2_PLEVEL_IN2SCNDRY_IN_cdc_to`

**Comment by AlvinRolling** [comment]
@stefanct Sorry, I just realized that I didn't address this problem a couple of days ago. The `synth_design` command still doesn't work. Anyway, I think the **OBUF** problem could be addressed using the way I mentioned above. I finished the implementation and programmed it on my FPGA, but have not tested it yet.

**Comment by stefanct** [comment]
@AlvinRolling yes, I think so too and have hopefully implemented something that works too. I will test that in about 12 hours.

**Comment by stefanct** [comment]
Unfortunately, the Linux kernel that I also built with the new(er) version of Xilinx' tools does hang immediately after getting control from uboot :(

    Copying Linux from SD to RAM...
    reading uImage
    3758680 bytes read in 380 ms (9.4 MiB/s)
    reading devicetree.dtb
    10276 bytes read in 19 ms (527.3 KiB/s)
    ## Booting kernel from Legacy Image at 02080000 ...
       Image Name:   Linux-4.14.0-xilinx
       Image Type:   ARM Linux Kernel Image (uncompressed)
       Data Size:    3758616 Bytes = 3.6 MiB
       Load Address: 00008000
       Entry Point:  00008000
       Verifying Checksum ... OK
    ## Flattened Device Tree blob at 02000000
       Booting using the fdt blob at 0x2000000
       Loading Kernel Image ... OK
       Loading Device Tree to 1ed1e000, end 1ed23823 ... OK

    Starting kernel ...

That's with the one from `tags/xilinx-v2018.3`... maybe my changes to `fpga/sw/kernel.config` are not correct. I created it basically by running `make ARCH=arm CROSS_COMPILE=arm-linux-gnueabihf-  olddefconfig`. This is rather different from the defconfig of the Xilinx's linux tree (`fpga/sw/linux-xlnx/arch/arm/configs/xilinx_zynq_defconfig`). However, a kernel built with `xilinx_zynq_defconfig` does not boot either!? :/

**Comment by stefanct** [comment]
The problems mentioned above were due to a bug in Xilinx UART driver (who would have thought...). After fixing that and a bunch of other problems related to the DT etc. I have now a pulpino version that can be successfully built for the Zedboard with Vivado 2015.1 *AND* 2018.3 by just changing an environment variable. Supporting other versions should also be possible with rather few changes to the build system (depending on how many spokes xilinx puts into your wheel). Supporting anything >2019.1 won't be easy due to the discontinuation of the SDK though.

One important change I made is that the SPI device has changed to `spidev1.0` no matter which Vivado version is used. The applications within the Pulpino repository have been changed and there is also a commit for the debug bridge (https://github.com/stefanct/debug_bridge/commit/d2730bd9e3b5c82aaaa16c7e930f84381689bdd4).

The current head of my patch set (which consists of 54 patches as of now...) is https://github.com/stefanct/pulpino/commit/e95dc276156c00ab7ce6c5edba18be8d110c3844
If anybody from the ETHZ (or bologna) want's to merge or review this I can open a PR if need be of course...
